src/main/java/io/confluent/connect/elasticsearch/bulk/BulkProcessor.java

```diff
@@ -122,20 +125,35 @@ public Thread newThread(Runnable r) {
     };
   }
 
-  private Runnable farmerTask() {
-    return new Runnable() {
-      @Override
-      public void run() {
-        log.debug("Starting farmer task");
-        try {
-          while (!stopRequested) {
-            submitBatchWhenReady();
+  // visible for testing
+  Runnable farmerTask() {
+    return () -> {
+      log.debug("Starting farmer task");
+      try {
+        List<Future<BulkResponse>> futures = new ArrayList<>();
+        while (!stopRequested) {
+          // submitBatchWhenReady waits for lingerMs so we won't spin here unnecessarily
+          futures.add(submitBatchWhenReady());
+
+          // after we submit, look at any previous futures that were completed and call get() on
+          // them so that exceptions are propagated.
+          List<Future<BulkResponse>> unfinishedFutures = new ArrayList<>();
+          for (Future<BulkResponse> f : futures) {
+            if (f.isDone()) {
+              log.debug("Calling get on completed future");
+              f.get();
+            } else {
+              unfinishedFutures.add(f);
+            }
```

Comment:
should we just (submit all and) wait for the first one with a timeout? otherwise this goes into a tight loop, right?

Reply:
yeah, the logic here is a bit tricky. It takes into account both cases `unsentRecords>batchSize` and `unsentRecords<batchSize`. For the former case, we enter into a tight loop creating Futures, each sending request of size batchSize. For the later, we wait for `lingerMs` to see how much of the batch we can fill before sending it off.

I'd kind of prefer not to change this and focus on fixing the bug in exception propagation.

Reply:
just curious, why did the exception handler not catch these errors already? looks like we are already applying it to the `farmer` as well as every thread in the `executor`, so any error in any of these threads should be automatically propagated to that handler, which in turn should fail the task.

Reply:
The uncaught exception handler wouldn't be invoked here because with `submit` the exception is considered part of the future output and wrapped in `ExcecutionException` on `get`. This is different than if you use the `execute` method on `ExecutorService`. With `execute`, the exception would be caught and handled by the unhandled exception handler as expected. This change will have the `farmer` thread handle and rethrow the `ExecutionException`s from BulkProcessor threads. Since the `farmer` thread is started directly, the rethrown exception would get handled by the uncaught exception handler as expected.

Reply:
Better explanation here - https://stackoverflow.com/a/29050729/12414477

We did not catch this earlier because our tests are calling `get` on the returned future and blocking on the response while our production code was not.

Reply:
sounds good. thanks, @ncliang!